plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -187,76 +236,98 @@ instance NFData MinimalImports
 
 type instance RuleResult MinimalImports = MinimalImportsResult
 
-newtype MinimalImportsResult = MinimalImportsResult
-  {getMinimalImportsResult :: [(LImportDecl GhcRn, Maybe T.Text)]}
+data MinimalImportsResult = MinimalImportsResult
+  { -- |For providing the code lenses we need to have a range, and a unique id
+    -- that is later resolved to the new text for each import. It is stored in
+    -- a list, because we always need to provide all the code lens in a file.
+    forLens        :: [(Range, Int)]
+    -- |For the code actions we have the same data as for the code lenses, but
+    -- we store it in a RangeMap, because that allows us to filter on a specific
+    -- range with better performance, and code actions are almost always only
+    -- requested for a specific range
+  , forCodeActions :: RM.RangeMap (Range, Int)
+    -- |For resolve we have an intMap where for every previously provided unique id
+    -- we provide a textEdit to allow our code actions or code lens to be resolved
+  , forResolve     :: IM.IntMap TextEdit }
 
 instance Show MinimalImportsResult where show _ = "<minimalImportsResult>"
 
 instance NFData MinimalImportsResult where rnf = rwhnf
 
+data EIResolveData = ResolveOne
+                      { uri      :: Uri
+                      , importId :: Int }
+                    | ResolveAll
+                      { uri :: Uri }
+                    deriving (Generic, A.ToJSON, FromJSON)
+
 exportedModuleStrings :: ParsedModule -> [String]
 exportedModuleStrings ParsedModule{pm_parsed_source = L _ HsModule{..}}
   | Just export <- hsmodExports,
     exports <- unLoc export
     = map (T.unpack . printOutputable) exports
 exportedModuleStrings _ = []
 
-minimalImportsRule :: Recorder (WithPriority Log) -> Rules ()
-minimalImportsRule recorder = define (cmapWithPrio LogShake recorder) $ \MinimalImports nfp -> do
+minimalImportsRule :: Recorder (WithPriority Log) -> (ModuleName -> Bool) -> Rules ()
+minimalImportsRule recorder modFilter = defineNoDiagnostics (cmapWithPrio LogShake recorder) $ \MinimalImports nfp -> runMaybeT $ do
   -- Get the typechecking artifacts from the module
-  tmr <- use TypeCheck nfp
+  (tmr, tmrpm) <- MaybeT $ useWithStale TypeCheck nfp
   -- We also need a GHC session with all the dependencies
-  hsc <- use GhcSessionDeps nfp
+  (hsc, _) <- MaybeT $ useWithStale GhcSessionDeps nfp
```

Comment:
ugh, this is a bit awkward because we now have two position mappings. I _think_ this is actually fine because we're never using any positions from this result. But it's worth a comment!

Reply:
I think hsc (GhcSessionDeps) is only used once to filter out exported modules, most of the heavy lifting is done by the (tmr)  (TypeCheck), so it shouldn't be a problem.